### DISCUSSION TITLE
How can I use datamodule in inference code?

### DISCUSSION THREAD

**Discussion by unolife** [manujosephv/pytorch_tabular#401] [Q&A]
@manujosephv , Idk why I can't add a comment on this issue https://github.com/manujosephv/pytorch_tabular/issues/396 

here is my fixed code to activate softmax

```
model = torch.load('./model.pt')
model.eval()
continuous_cols = df.select_dtypes(include=['float64', 'int64']).columns.tolist()
categorical_cols = df.select_dtypes(include=['object', 'category']).columns.tolist()
x = {
    "continuous": torch.tensor(df[continuous_cols].values, dtype=torch.float32), 
    "categorical": torch.tensor(df[categorical_cols].values, dtype=torch.float32)
    }
logits = model.predict(x)
res = F.softmax(logits, dim=1)
```
But Idk how can I apply datamodule in this code, can you give me some advice please??

**Comment by manujosephv** [comment]
so, instead of constructing a dictionary directly from the dataframe, you should be using the [`datamodule.prepare_inference_dataloader`](https://pytorch-tabular.readthedocs.io/en/latest/apidocs_coreclasses/#pytorch_tabular.TabularDatamodule.prepare_inference_dataloader) to create a dataloader with desired batch size, and then iterate over that dataloader to do predictions.
Here is the simplified code which is used in the `TabularModel.predict`

```python
point_predictions = []
for batch in progress_bar(inference_dataloader):
    for k, v in batch.items():
        if isinstance(v, list) and (len(v) == 0):
            continue  # Skipping empty list
        batch[k] = v.to(model.device)
        y_hat=F.softmax(model.predict(batch), dim=1)
        point_predictions.append(y_hat.detach().cpu())
point_predictions = torch.cat(point_predictions, dim=0)
```

**Reply by manujosephv** [reply]
Also, I've copied these comments on to #396  as well
